### PR TITLE
Reorganize docs v2

### PR DESCRIPTION
**Pull request by ichard26**
Resolves GH-1759.

---

I'm finally done :tada: 

To anyone trying to review this, I recommend looking at the rendered previews I've provided below OR doing this commit-by-commit (looking at the whole diff will be painful). I've tried to provide a decent commit history but yeah, my apologies. Take your time and don't worry about asking for a lot of changes. I want this structure to last us a long time. 

**I'd ideally want as many reviews as possible but as a minimum I'd like two maintainer and one user reviews.**

What I'm looking for in terms of feedback and reviews:

- Does the structure make sense? 
  - Can you navigate it and not end up confused?
- Is there any important information I'm missing (I do plan to write a bunch of new docs after I'm done with this but it's good to know early just in case some structural changes have to happen)
- Do you think this structure is expandable and flexible enough? (i.e. will it last us a _long_ time?)

--- 

- Look at the reorganized docs here: https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/
- Look at the rewritten README here: https://github.com/ichard26/black/tree/reorganize-docs-v2#readme
- Look at the rewritten CONTRIBUTING here: https://github.com/ichard26/black/blob/reorganize-docs-v2/CONTRIBUTING.md

---

**FYI, I want to handle merging this myself since I'd like to edit the commit message to my liking.**

**Also I'll need this to be merged as close as possible before a release since the README.md / CONTRIBUTING.md links are dependent on the stable version on RTD having these changes at the same time.**

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Thanks so much! I'm going to take this occasion to just read the whole docs and comment on anything that seems off.

* https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/the_black_code_style/current_style.html
  * "It doesn’t take previous formatting into account." – except for the magic trailing comma, and preserving newlines
  * "As an experimental option, Black splits long strings" – it's strange that there's all this text *before* we mention that this only happens if the experimental flag is on. I'd recommend moving the name of the option to the beginning of the paragraph.
  * "the parenthesis can also be successfully omitted" – should be *parentheses", you can never safely omit a single parenthesis :)
* https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/the_black_code_style/future_style.html
   * Could also mention that as of a recent version of Python you can parenthesize withs instead. If I recall correctly this actually works as of 3.9, but was undocumented, and it will be documented in 3.10.
* https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/getting_started.html
   * "Although Python 2 support needs the typed_ast dependency which be installed with pip install black[python2]." – remove "Although", add a comma before "which"
 * https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/usage_and_configuration/the_basics.html
   * "Black doesn’t provide many options" – only 20 of them. I don't think we can make this claim with a straight face any more.
 * https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/usage_and_configuration/file_collection_and_discovery.html
    * Should mention gitignore support
 * https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/integrations/index.html
   *  "Any tool that can pipe code through Black" – remove "that". Also, in the next sentence put `--check` in backticks.
 * https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/integrations/github_actions.html
   * Should explain what this action does, something like "This action will check that your code matches the *Black* code style."
 * https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/guides/using_black_with_other_tools.html
    *  "Although, this style is PEP 8 compliant" – remove the comma
 * https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/contributing/the_basics.html
    * "I am using macOS myself and CentOS at work." – I guess this is what Lukasz was using a few years ago. Let's just remove this sentence.
    * "If you feel this PR not require a changelog entry" – needs a "does" somewhere
 * https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/change_log.html
    * It'd be nice if the PR numbers autolinked to the PR on GitHub

Those are mostly pretty minor. Overall this makes the docs a lot better and I hope we'll get a release with these reorganized docs soon!

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -24,8 +24,12 @@ repos:
...
 
-  - repo: https://github.com/pre-commit/mirrors-prettier
-    rev: v2.2.1
+  - repo: https://github.com/executablebooks/mdformat
+    rev: 0.7.5
     hooks:
-      - id: prettier
-        args: [--prose-wrap=always, --print-width=88]
+      - id: mdformat
+        args: [--number, --wrap=88]
```
Comment (hukkinj1) [inline .pre-commit-config.yaml:31]: I'd argue that removing `--number` arguement would be more in the spirit of Black (minizing diffs is great, [docs](https://mdformat.readthedocs.io/en/stable/users/style.html#ordered-lists) explain this).

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -24,8 +24,12 @@ repos:
...
+        additional_dependencies:
+          - mdformat-gfm
+          - mdformat-myst
+        exclude: ^.github/ISSUE_TEMPLATE
```
Comment (hukkinj1) [inline .pre-commit-config.yaml:35]: It should be safe to format these as long as `frontmatter` syntax support is enabled (which it is because `mdformat-myst` includes it).

You were worried about safety in the commit message. Note that mdformat has a check to ensure AST stays unchanged (similar to Black), which Prettier to my knowledge doesn't have.

**Comment by felix-hilden** [comment]
Would be my pleasure! First of all, it all looks so good. Worlds apart with the current structure 👌

Jelle commented on quite a few points already:

>       * "Black doesn’t provide many options" – only 20 of them. I don't think we can make this claim with a straight face any more.

Perhaps the intent could still be communicated to the users. Something like: "Black is opinionated, so style configuration options are deliberately limited and rarely added".

>     * https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/change_log.html
>       
>       * It'd be nice if the PR numbers autolinked to the PR on GitHub

See #2087 and maybe discuss there? Doesn't necessarily have to be RST!

- [Getting started](https://ichard26-testblackdocs.readthedocs.io/en/reorganize-docs-v2/getting_started.html): "Installing from GitHub could be removed, as the section is very short and right after "Installation".
- Getting started: "Next steps" has many sentences in individual paragraphs. That's pretty jarring to read. Have them in a single paragraph?

Having the contributing docs online is a great change too. In line with that, should the change log be single sourced as well? Now the readme links to the GitHub tree, but it could be changed to a RTD link (because the RTD `latest` has changes from master I presume). And even further: the readme duplicates a *lot* of information on the documentation. Does it need to? If users are directed to RTD from everywhere, the GitHub readme can serve contributors and the *repository* more, not users.

I'm not sure if all of these changes are appropriate for this PR 😅 I agree: this is great, and after fixing the small details it surely is good to go as an incremental improvement (read: giant leap for mankind). All the rest can be broken down into future contributions.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (ichard26) [reply .pre-commit-config.yaml:35]: Alright you've convinced me :slightly_smiling_face: You should definitely advertise the AST safety feature of mdformat IMO (why not mention it in your "why not prettier" section).

**Comment by ichard26** [comment]
Haha, I've been reading over my own work and I see a bunch of little errors too :)

I tried to avoid writing too much new content / editing existing content (other than the Getting Started page plus all of the section index pages) so this wouldn't be as painful to review. So IMO unless the suggestion falls under something I (re)wrote, I'll add them in a separate PR. 

---

> Perhaps the intent could still be communicated to the users. Something like: "Black is opinionated, so style configuration options are deliberately limited and rarely added".

:+1:

> Getting started: "Installing from GitHub could be removed, as the section is very short and right after "Installation".

IDC but sure.

> Getting started: "Next steps" has many sentences in individual paragraphs. That's pretty jarring to read. Have them in a single paragraph?

:+1: 

> In line with that, should the change log be single sourced as well? 

:+1: 

> the readme duplicates a lot of information on the documentation. Does it need to? 

We've already tried hard to get rid as much duplicated content as possible, but I didn't think I would be able to get approval to strip 98% of the README content :upside_down_face: But in general I agree it shouldn't and I'll take another look seeing what else I should remove.

**Comment by felix-hilden** [comment]
Yeah there's definitely an argument to be had about not simply having a "users go to RTD" and nothing else. I like it that way, but I imagine there are lots of people who prefer to have something resembling what's currently there. And again, incremental changes :D

**Comment by wbolster** [comment]
quick feedback

the editor integration for emacs should point to python-black.el at https://github.com/wbolster/emacs-python-black

the link to "reformatter.el" should go, since it's a low level library to build integrations, and python-black.el uses it.

🙏

**Comment by ichard26** [comment]
@wbolster noted, I'll take of it soon (or eventually). Thank you for noticing and reporting!

File: .pre-commit-config.yaml (hunk 1 above)
Comment (ichard26) [reply .pre-commit-config.yaml:31]: I literally copied and pasted the configuration you suggested in that other issue but OK :laughing: . Looking into this, I agree omitting the `--number` argument is better.

Comment (hukkinj1) [reply .pre-commit-config.yaml:31]: Haha yeah I know. The initial suggestion was just meant to be as close to the Prettier style that was in use as possible. I thought you may not want to change the style in the scope of this PR. But seeing that the diffs are pretty huge anyways, maybe you do :smile: ?

Comment (hukkinj1) [reply .pre-commit-config.yaml:31]: In fact, you may want to consider omitting `--wrap=88` for the same reason and adopt [semantic line breaks](https://sembr.org/). That change however requires a lot of manual effort unless you migrate to it gradually.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by ichard26** [comment]
OK, updated the branch and incorporated all of the feedback I've received so far. I know I said I wanted to do these after landing this but given there's going to be no time between this being merged and a release getting pushed, I want these changes to make it in.

@cooperlees I heard you wanted to do a review, so count this as your reminder to so, thank you in advance! Any other final reviews or comments are also welcomed.

Finally a reminder to fellow maintainers that **I want to handle merging this myself since I'd like to edit the commit message to my liking.**

File: docs/conf.py [hunk 3]
```diff
@@ -341,20 +147,20 @@ def process_sections(
...
+#     The font size ('10pt', '11pt' or '12pt').
+#
+#     'pointsize': '10pt',
+#     Additional stuff for the LaTeX preamble.
+#
+#     'preamble': '',
+#     Latex figure (float) alignment
+#
+#     'figure_align': 'htbp',
+# }
```
Comment (cooperlees) [inline docs/conf.py:163]: Can we delete this?

File: docs/conf.py [hunk 4]
```diff
@@ -78,131 +29,10 @@ def make_pypi_svg(version: str) -> None:
...
-            f.write(contents)
-        processed_sections[section.name] = section
-        modified_files.add(target_path)
-
-
 # -- Project information -----------------------------------------------------
 
 project = "Black"
-copyright = "2020, Łukasz Langa and contributors to Black"
+copyright = "2021, Łukasz Langa and contributors to Black"
```
Comment (cooperlees) [inline docs/conf.py:35]: Should we just do a from 2018 to Present or something ?

File: docs/contributing/gauging_changes.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
+# Gauging changes
+
+A lot of the time, your change will affect formatting and/or performance. Quantifying
+these changes is hard, so we have tooling to help make it easier.
+
+It's recommended you take a look at quantifable changes your change will cause before
+submitting a PR if it seems disruptive enough.
```
Comment (cooperlees) [inline docs/contributing/gauging_changes.md:7]: I found this hard to parse.
```suggestion
It's recommended you evaluate the quantifable changes your black formatting change
causes before submitting a PR. Think about if the change seems disruptive enough 
to cause frustration to projects that are already "black formatted".
```

File: docs/contributing/gauging_changes.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+## black-primer
+
+`black-primer` is a tool built for CI (and humans) to have _Black_ `--check` a number of
+(configured in `primer.json`) Git accessible projects in parallel. _(A PR will be
+accepted to add Mercurial support.)_
```
Comment (cooperlees) [inline docs/contributing/gauging_changes.md:13]:
```suggestion
`black-primer` is a tool built for CI (and humans) to have _Black_ `--check` a number of
Git accessible projects in parallel (configured in `primer.json`). _(A PR will be
accepted to add Mercurial support.)_
```

**Review by hukkinj1 — COMMENTED** [review]
>Prettier works fine for all of the default MyST syntax

@ichard26 I can't force you to use mdformat :smile: but will mention that this is not true.

A very simple example doc:

```markdown
a line
% a comment
another line
```
Prettier output:
```markdown
a line % a comment another line
```

File: docs/index.rst [hunk 7]
```diff
@@ -16,10 +16,24 @@ can focus on the content instead.
...
-
-   `Black is beta <installation_and_usage.html#note-this-is-a-beta-product>`_.
+.. admonition:: Note - this is a beta product
+
+   *Black* is already `successfully used <https://github.com/psf/black#used-by>`_ by
+   many projects, small and big. It also sports a decent test suite. However, it is
+   still very new. Things will probably be wonky for a while. This is made explicit by
+   the "Beta" trove classifier, as well as by the "b" in the version number. What this
+   means for you is that **until the formatter becomes stable, you should expect some
+   formatting to change in the future**. That being said, no drastic stylistic changes
```
Comment (cooperlees) [inline docs/index.rst:26]: Not to fussed here, take it of leave it. I just don't like a ~3 year old project being called new.
```suggestion
   many projects, small and big. Black has a comprehensive test suite, with efficient parallel tests, our own auto  formatting and parallel Continuous Integration runner. However, _Black_ is still beta. Things will probably be wonky for a while. This is made explicit by
   the "Beta" trove classifier, as well as by the "b" in the version number. What this
   means for you is that **until the formatter becomes stable, you should expect some
   formatting to change in the future**. That being said, no drastic stylistic changes
```

File: docs/integrations/github_actions.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# GitHub Actions integration
+
+You can use _Black_ within a GitHub Actions workflow without setting your own Python
+environment. Great for enforcing that your code matches the _Black_ code style.
```
Comment (cooperlees) [inline docs/integrations/github_actions.md:4]:
```suggestion
environment. Great for enforcing that your code matches the _Black_ code style.
We recommend the use of the `@stable` tag, but per version tags also exist if you prefer that.
```

File: docs/integrations/source_version_control.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# Version control integration
+
+Use [pre-commit](https://pre-commit.com/). Once you
+[have it installed](https://pre-commit.com/#install), add this to the
+`.pre-commit-config.yaml` in your repository:
+
+```yaml
+repos:
+  - repo: https://github.com/psf/black
+    rev: 20.8b1 # Replace by any tag/version: https://github.com/psf/black/tags
```
Comment (cooperlees) [inline docs/integrations/source_version_control.md:10]: People will copy paste
```suggestion
    rev: stable  # Replace by any tag/version: https://github.com/psf/black/tags
```

**Comment by ichard26** [comment]
@hukkinj1 well for that we'll use `<! -- {content} -->`. For our purposes, it works fine (and while stuff inside roles, directives, etc. aren't formatted by prettier, they're not broken by it). We can always revisit it. My apologies for throwing away your work like that. Thanks for all your help!

File: docs/the_black_code_style/future_style.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+```py3
+with \
+     make_context_manager(1) as cm1, \
+     make_context_manager(2) as cm2, \
+     make_context_manager(3) as cm3, \
+     make_context_manager(4) as cm4 \
+:
+    ...  # backslashes and an ugly stranded colon
+```
```
Comment (cooperlees) [inline docs/the_black_code_style/future_style.md:32]: I don't know if this is still true - I get code like this these days: https://github.com/cooperlees/json2netns/blob/main/src/json2netns/tests/netns.py#L43-L45

I guess maybe cause I don't have the '\'

File: docs/usage_and_configuration/the_basics.md [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+                                    2>/dev/null.
+
+    -v, --verbose                   Also emit messages to stderr about files
+                                    that were not changed or were ignored due to
+                                    exclusion patterns.
+
+    --version                       Show the version and exit.
+    --config FILE                   Read configuration from FILE path.
+    -h, --help                      Show this message and exit.
+```
```
Comment (cooperlees) [inline docs/usage_and_configuration/the_basics.md:123]: Side note: I know it's nice, but I hate how these go out of date (And do for other projects). I've wondered if we can in a follow up PR create a CI action to dump `--help` to a file and include it somehow?

Comment (cooperlees) [reply docs/usage_and_configuration/the_basics.md:123]: Filed https://github.com/pypa/bandersnatch/issues/907 for bandersnatch to see if we can POC something there ...

File: docs/usage_and_configuration/the_basics.md [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+## Configuration via a file
+
+_Black_ is able to read project-specific default values for its command line options
+from a `pyproject.toml` file. This is especially useful for specifying custom
+`--include` and `--exclude`/`--force-exclude`/`--extend-exclude` patterns for your
+project.
+
+**Pro-tip**: If you're asking yourself "Do I need to configure anything?" the answer is
+"No". _Black_ is all about sensible defaults.
```
Comment (cooperlees) [inline docs/usage_and_configuration/the_basics.md:135]:
```suggestion
"No". _Black_ is all about sensible defaults. Applying those defaults will have your code in compliance with many other _Black_ formatted projects.
```

**Review by cooperlees — APPROVED** [review]
LGTM. I read closely and found a few things hard to parse so offered suggestions. Take or leave them. Then lets get this sucker merged so we can release awesome docs + refactor ready for PyCon and possible a bug fix or two :D

File: docs/conf.py (hunk 3 above)
Comment (ichard26) [reply docs/conf.py:163]: Sure, this might be useful for building PDF documentation but I have no experience with that so that will have to wait for another day.

File: docs/the_black_code_style/future_style.md (hunk 10 above)
Comment (ichard26) [reply docs/the_black_code_style/future_style.md:32]: I'm confused, the style described in this document is planned, but hasn't be implemented yet. This used to exist the main style doc but that was misleading and confusing.

File: docs/contributing/gauging_changes.md (hunk 5 above)
Comment (ichard26) [reply docs/contributing/gauging_changes.md:7]: I tend to either write too densely or with too many words, this seems wayyy more reasonable. Thanks!

**Comment by hukkinj1** [comment]
> @hukkinj1 well for that we'll use <! -- {content} --> ... so okay maybe not all of the default syntax is supported but.... for our purposes, it works fine (and while stuff inside roles, directives, etc. aren't formatted by prettier, they're not broken by it). We can always revisit it. My apologies for throwing away your work like that. Thanks for all your help!

No worries. Note that you can disable certain myst syntax in sphinx conf.py. Example here https://github.com/executablebooks/mdformat/blob/ba3385a3e5c08cbf8404b3fb223660047baab2d5/docs/conf.py#L64

Comments, block level math and blockbreaks at least will be broken by prettier, so might be a good idea to disable those syntaxes.

File: docs/contributing/gauging_changes.md (hunk 5 above)
Comment (JelleZijlstra) [reply docs/contributing/gauging_changes.md:7]: Also quantifable -> quantifiable

**Review by ambv — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
🎉
